python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java

```diff
@@ -581,6 +584,71 @@ def foo3(a: int = "123"):
     Assertions.assertThat(lastType.unwrappedType()).isEqualTo(INT_TYPE);
   }
 
+  @Test
+  void inferFunctionParameterTypes() {
+    FileInput root = inferTypes("""
+      def foo(param: int, *args, **kwargs):
+        ...
+      """);
+
+    var functionDef = (FunctionDef) root.statements().statements().get(0);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(0).declaredType().unwrappedType()).isEqualTo(INT_TYPE);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(1).declaredType().unwrappedType()).isEqualTo(TUPLE_TYPE);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(2).declaredType().unwrappedType()).isEqualTo(DICT_TYPE);
+  }
+
+  @Test
+  void inferFunctionParameterTypes2() {
```

Comment:
I'm not sure I understand this test. I guess `param: A` is unknown because the class `A` is defined twice, and only the trivial type inference has run at this point. Is that correct?

Do we need to have the `foo` methods in `A` for that behavior to show (if not, I would suggest to remove them - especially as their name clashes with the function `foo`, which adds confusion).